File: next-js-boilerplate/src/api/server/premium/growth-stats.ts
import { apiFetch } from "@/lib/api-client";
import { GQL_URL } from "@/constants/api/urls";
import { POST } from "@/constants/api/methods";
import { JSON_CONTENT_TYPE_HEADER } from "@/constants/api/headers";

export interface GrowthStats {
  totalUsers: number;
  newUsersLast7Days: number;
  totalPosts: number;
  totalFriendships: number;
}

export async function fetchGrowthStatsServer(): Promise<GrowthStats> {
  const res = await apiFetch(GQL_URL, {
    method: POST,
    headers: JSON_CONTENT_TYPE_HEADER,
    body: JSON.stringify({
      query: `query { growthStats { totalUsers newUsersLast7Days totalPosts totalFriendships } }`,
    }),
  });
  if (!res.ok) throw new Error("Failed to fetch growth stats");
  const data = await res.json();
  // A GraphQL error rides inside a 200 response, so `res.ok` alone doesn't
  // catch it — this previously fell through to `?? null`, which the caller
  // (loadPremiumGrowthStats) couldn't distinguish from "loading", leaving the
  // skeleton spinner up forever with no error toast.
  if (data.errors || !data.data?.growthStats) {
    throw new Error("Failed to fetch growth stats");
  }
  return data.data.growthStats;
}

File: flutter-boilerplate/lib/api/server/premium/growth_stats.dart
import 'package:dio/dio.dart';
import 'package:flutter_boilerplate/lib/api_client.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class GrowthStats {
  final int newUsersLast7Days;
  final int totalPosts;
  final int totalFriendships;

  const GrowthStats({
    required this.newUsersLast7Days,
    required this.totalPosts,
    required this.totalFriendships,
  });

  factory GrowthStats.fromJson(Map<String, dynamic> json) {
    return GrowthStats(
      newUsersLast7Days: json['newUsersLast7Days'] as int,
      totalPosts: json['totalPosts'] as int,
      totalFriendships: json['totalFriendships'] as int,
    );
  }
}

final growthStatsServerProvider =
    Provider((ref) => GrowthStatsServer(ref.read(dioProvider)));

const _query =
    'query GrowthStats { growthStats { totalUsers newUsersLast7Days totalPosts totalFriendships } }';

class GrowthStatsServer {
  final Dio _dio;

  GrowthStatsServer(this._dio);

  Future<GrowthStats> call() async {
    final response = await _dio.post<dynamic>(
      '/graphql',
      data: {'query': _query},
    );
    final body = response.data as Map<String, dynamic>;
    if (body['errors'] != null) {
      throw DioException(
        requestOptions: response.requestOptions,
        message: 'Failed to fetch growth stats',
      );
    }
    final result = (body['data'] as Map<String, dynamic>)['growthStats']
        as Map<String, dynamic>;
    return GrowthStats.fromJson(result);
  }
}
